🌱 Refactor clusterctl alpha rollout

hiromu-a5a opened the pull request:
<!-- please add an icon to the title of this PR (see https://sigs.k8s.io/cluster-api/CONTRIBUTING.md#contributing-a-patch), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

**What this PR does / why we need it**:

A tiny refactoring of the RolloutOptions structure used by clusterctl alpha rollout. Defined RolloutOptions for each command.

The current codes for rollout reuses a common RolloutOptions structure among all kinds of rollout commands, i.e., undo, resume, restart and pause. However, the required inputs for those commands might be different in the future extension. From this perspective, the structure should be defined separately for each command.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Tracking Issue:
https://github.com/kubernetes-sigs/cluster-api/issues/3439

hiromu-a5a commented:
Following the other `cluster alpha rollout` patches, I didn't create an specific issue for this PR also because it's a small PR. If I have to create an issue, please kindly let me know.

fabriziopandini commented:
/ok-to-test

fabriziopandini reviewed (commented):
I'm generally +1 for the change, but let's wait for a pass from clusterctl reviewers first.

Jont828 reviewed (commented):
Thanks for the contribution! I think it makes sense to make the roll out options more fine-grained. 

/lgtm

Jont828 commented:
Only thing seems to be that the main test is failing btw.

fabriziopandini commented:
/retest
seems a flake

fabriziopandini commented:
la st nits from my side assuming test failures are flakes.
Given that this PR changes our API let's use ⚠️ instead of 🌱  in the title.
Also, let's add a note for this change in https://main.cluster-api.sigs.k8s.io/developer/providers/v1.3-to-v1.4.html#api-changes

hiromu-a5a commented:
Thank you for your comments.

Regarding pull-cluster-api-apidiff-main

> The test module, clusterctl, and experiments do not provide any backward compatible guarantees.

hiromu-a5a commented:
Thank you for your prompt reply! 

I added note.

I have one question. The failing test `pull-cluster-api-apidiff-main` shows the changes are incompatible as follows. Any appropriate ways to solve this? 
According to [the past discussion](https://kubernetes.slack.com/archives/C8TSNPY4T/p1661858300361619?thread_ts=1661857399.179689&cid=C8TSNPY4T) and [the Contributing Guideline](https://github.com/kubernetes-sigs/cluster-api/blob/main/CONTRIBUTING.md#codebase-and-go-modules), clusterctl is not required to provide any backward compatible guarantees but should be checked by approvers/maintainers if the change is acceptable.

```
 Incompatible changes:
  - (*clusterctlClient).RolloutRestart: changed from func(RolloutOptions) error to func(RolloutRestartOptions) error
  - (*clusterctlClient).RolloutResume: changed from func(RolloutOptions) error to func(RolloutResumeOptions) error
  - (*clusterctlClient).RolloutUndo: changed from func(RolloutOptions) error to func(RolloutUndoOptions) error
  - AlphaClient.RolloutRestart: changed from func(RolloutOptions) error to func(RolloutRestartOptions) error
  - AlphaClient.RolloutResume: changed from func(RolloutOptions) error to func(RolloutResumeOptions) error
  - AlphaClient.RolloutUndo: changed from func(RolloutOptions) error to func(RolloutUndoOptions) error
```

killianmuldoon reviewed (approved):
> I have one question. The failing test pull-cluster-api-apidiff-main shows the changes are incompatible as follows. Any appropriate ways to solve this?

You're right that the guarantees for clusterclt are different :slightly_smiling_face:. The test will fail, but that won't block this PR getting merged. In this case it acts more as a signal to reviewers to consider what is being changed in the API.

/lgtm

fabriziopandini reviewed (commented):
/lgtm
/approve

fabriziopandini commented:
/hold
checking the lint job

fabriziopandini commented:
/hold cancel
